From c3385c499da1af65f4050646f961b1ac66112f20 Mon Sep 17 00:00:00 2001
From: Lars Schmertmann <Lars.Schmertmann@governikus.de>
Date: Mon, 9 Jun 2025 12:32:11 +0200
Subject: Android: Unify behavior of Accessible.description with other
 platforms

On Windows, macOS and iOS both text and description are read out. On
Android the description is only used, when there is no name/text and
the result is used in AccessibilityNodeInfo::setContentDescription
while AccessibilityNodeInfo::setText is ignored. Using the content
description only is right, because Android will prioritize description
over text. So we need to do our own concatenation to use both.
Using ", " as separator looks like the right way, because Android use
it to separate the text from the role too. This was checked by enabling
TalkBack settings -> Advanced settings -> Developer settings -> Display
speech output.

Fixes: QTBUG-128494
Pick-to: 6.10
Change-Id: Ib25e993ffc8614b9c9ccdc37207f2c3c496ecb1f
Reviewed-by: Assam Boudjelthia <assam.boudjelthia@qt.io>
(cherry picked from commit ce101c2c3c04f5ccd46abf7024662d0bc4e5a3a3)
---
 src/plugins/platforms/android/androidjniaccessibility.cpp | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
index 73304a9f20178129d966f8a8af48a0f6f5c7f020..4795fffe3143f33e5fc2e142cb67cda55de159c4 100644
--- x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
+++ y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
@@ -559,8 +559,12 @@ namespace QtAndroidAccessibility
         if (iface && iface->isValid()) {
             bool hasValue = false;
             desc = iface->text(QAccessible::Name);
-            if (desc.isEmpty())
-                desc = iface->text(QAccessible::Description);
+            const QString descStr = iface->text(QAccessible::Description);
+            if (!descStr.isEmpty()) {
+                if (!desc.isEmpty())
+                    desc.append(QStringLiteral(", "));
+                desc.append(descStr);
+            }
             if (desc.isEmpty()) {
                 desc = iface->text(QAccessible::Value);
                 hasValue = !desc.isEmpty();
